File: desktop-app/backend/app/optimization/fingerprints.py
"""Stable fingerprints for task equivalence vs execution configuration.

task_fingerprint: model-independent — proves same task across substitution.
execution_configuration_fingerprint: includes model/provider — proves how it ran.
baseline_evidence_hash: immutable baseline content identity for paid idempotency.
"""

from __future__ import annotations

import json
from typing import Any

from ..core.hashing import sha256_json, sha256_text
from ..db_models import Trace
from ..providers.models import LLMRequest


def _parse_meta(trace: Trace) -> dict[str, Any]:
    if not trace.metadata_json:
        return {}
    try:
        raw = json.loads(trace.metadata_json)
        return raw if isinstance(raw, dict) else {}
    except Exception:
        return {}


def task_fingerprint_from_trace(trace: Trace) -> str:
    """Model-independent task identity for baseline ↔ candidate equivalence."""
    snap_json = getattr(trace, "request_snapshot_json", None)
    if snap_json:
        try:
            from ..evidence.replay import ReplayableRequestSnapshot

            snap = ReplayableRequestSnapshot.model_validate_json(snap_json)
            return snap.task_fingerprint()
        except Exception:
            pass
    meta = _parse_meta(trace)
    payload = {
        "input_hash": sha256_text(trace.input_text) or sha256_text(trace.expected_output),
        "symbol": trace.symbol or "",
        "workflow": trace.workflow or "",
        "temperature": meta.get("temperature"),
        "system_prompt_version": meta.get("system_prompt_version") or meta.get("prompt_version"),
        "system_prompt_hash": meta.get("system_prompt_hash") or sha256_text(meta.get("system_prompt")),
        "response_format": meta.get("response_format"),
        "tool_schema_hash": meta.get("tool_schema_hash"),
        "context_id": meta.get("context_id") or meta.get("retrieval_context_id"),
        "model_config": meta.get("model_config") or meta.get("generation_config"),
        "prompt_config_version": meta.get("prompt_config_version") or meta.get("app_prompt_version"),
    }
    return sha256_json(payload)


def task_fingerprint_from_request(request: LLMRequest) -> str:
    """Canonical task identity from an LLMRequest (excludes provider/model)."""
    meta = dict(request.metadata or {})
    messages = []
    for msg in request.messages:
        messages.append(
            {
                "role": msg.role,
                "content": msg.content,
                "name": msg.name,
                "tool_call_id": msg.tool_call_id,
                # tool_calls on assistant messages affect task shape
                "tool_calls": [tc.model_dump() for tc in (msg.tool_calls or [])] or None,
            }
        )
    tools = None
    if request.tools:
        tools = [t.model_dump() for t in request.tools]
    payload = {
        "messages": messages,
        "tools": tools,
        "tool_choice": request.tool_choice if request.tools else None,
        "temperature": request.temperature,
        "max_tokens": request.max_tokens,
        "response_format": meta.get("response_format"),
        "context_id": meta.get("context_id") or meta.get("retrieval_context_id"),
        "workflow": meta.get("workflow"),
        "system_prompt_version": meta.get("system_prompt_version") or meta.get("prompt_version"),
        "prompt_config_version": meta.get("prompt_config_version") or meta.get("app_prompt_version"),
        "model_config": meta.get("model_config") or meta.get("generation_config"),
    }
    return sha256_json(payload)


def execution_configuration_fingerprint(
    *,
    provider: str | None,
    model: str | None,
    temperature: float | None,
    max_tokens: int | None,
    extra: dict[str, Any] | None = None,
) -> str:
    """How the candidate was executed (includes model substitution target)."""
    return sha256_json(
        {
            "provider": provider,
            "model": model,
            "temperature": temperature,
            "max_tokens": max_tokens,
            "extra": extra or {},
        }
    )


def baseline_evidence_hash(traces: list[Trace]) -> str:
    """Immutable baseline evidence identity for execution idempotency.

    Changing content/config hashes changes this key even if Trace IDs are reused.
    Timestamps are intentionally excluded.
    """
    items = []
    for t in sorted(traces, key=lambda x: x.id):
        meta = _parse_meta(t)
        items.append(
            {
                "trace_id": t.id,
                "request_id": t.request_id,
                "input_hash": t.input_hash or sha256_text(t.input_text),
                "output_hash": t.output_hash or sha256_text(t.output_text),
                "task_fingerprint": task_fingerprint_from_trace(t),
                "provider": t.provider,
                "model": t.model or t.requested_model,
                "symbol": t.symbol,
                "workflow": t.workflow,
                "protected": t.protected,
                "system_prompt_version": meta.get("system_prompt_version") or meta.get("prompt_version"),
                "tool_schema_hash": meta.get("tool_schema_hash"),
                "context_id": meta.get("context_id") or meta.get("retrieval_context_id"),
                "response_format": meta.get("response_format"),
            }
        )
    return sha256_json(items)

File: api/_lib/product/fingerprints.js
import { sha256Json, sha256Text } from './hashing.js';

function meta(trace) {
  return trace.metadata && typeof trace.metadata === 'object' ? trace.metadata : {};
}

/** Model-independent task identity for baseline ↔ candidate equivalence. */
export function taskFingerprint(trace) {
  const m = meta(trace);
  return sha256Json({
    input_hash: sha256Text(trace.input_text) || sha256Text(trace.expected_output),
    symbol: trace.symbol || '',
    workflow: trace.workflow || '',
    temperature: m.temperature ?? null,
    system_prompt_version: m.system_prompt_version || m.prompt_version || null,
    system_prompt_hash: trace.system_prompt_hash || m.system_prompt_hash || sha256Text(m.system_prompt) || null,
    prompt_shape_hash: sha256Json((trace.messages || []).map((x) => x.role)),
    response_format: m.response_format ?? null,
    tool_schema_hash: m.tool_schema_hash ?? null,
    context_id: m.context_id || m.retrieval_context_id || null,
    model_config: m.model_config || m.generation_config || null,
    prompt_config_version: m.prompt_config_version || m.app_prompt_version || null,
  });
}

/** Immutable baseline evidence identity for paid idempotency (timestamps excluded). */
export function baselineEvidenceHash(traces) {
  const items = [...traces]
    .sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0))
    .map((t) => {
      const m = meta(t);
      return {
        trace_id: t.id,
        request_id: t.request_id,
        input_hash: t.input_hash || sha256Text(t.input_text),
        output_hash: t.output_hash || sha256Text(t.output_text),
        task_fingerprint: taskFingerprint(t),
        provider: t.provider ?? null,
        model: t.model || t.requested_model || null,
        symbol: t.symbol ?? null,
        workflow: t.workflow ?? null,
        protected: Boolean(t.protected),
        system_prompt_version: m.system_prompt_version || m.prompt_version || null,
        system_prompt_hash: t.system_prompt_hash || m.system_prompt_hash || null,
        tool_schema_hash: m.tool_schema_hash ?? null,
        context_id: m.context_id || m.retrieval_context_id || null,
        response_format: m.response_format ?? null,
      };
    });
  return sha256Json(items);
}

/** Behaviour-affecting factors two traces must share to be reusable. */
export function reuseIdentity(trace) {
  const m = meta(trace);
  return {
    input_hash: sha256Text(trace.input_text) || sha256Text(trace.expected_output),
    symbol: trace.symbol || '',
    workflow: trace.workflow || '',
    provider: (trace.provider || '').toLowerCase(),
    model: (trace.model || trace.requested_model || '').trim(),
    temperature: m.temperature ?? null,
    system_prompt_version: m.system_prompt_version || m.prompt_version || null,
    system_prompt_hash: trace.system_prompt_hash || m.system_prompt_hash || sha256Text(m.system_prompt) || null,
    response_format: m.response_format ?? null,
    tool_schema_hash: m.tool_schema_hash ?? null,
    context_id: m.context_id || m.retrieval_context_id || null,
    model_config: m.model_config || m.generation_config || null,
  };
}

export function reuseIdentityHash(trace) {
  const identity = reuseIdentity(trace);
  if (!identity.input_hash) return null;
  return sha256Json(identity);
}

export function unsafeForReuse(trace) {
  if (trace.protected) return 'Protected/high-risk traces are not eligible for exact reuse.';
  const m = meta(trace);
  if (m.side_effects || m.non_idempotent || m.state_changing) return 'Trace metadata indicates non-idempotent or state-changing behavior.';
  if ((m.personalized || m.session_id) && !m.context_id && !m.retrieval_context_id) return 'Personalized/session-specific traces lack a safe context identity.';
  if (m.tools_executed || m.has_tool_calls) return 'Traces with tool execution are not eligible for exact reuse.';
  if (!trace.output_text) return 'Baseline output missing; cannot reuse an empty result.';
  if (!(trace.input_text || trace.expected_output)) return 'No normalized input/task identity available.';
  return null;
}
